python/kserve/kserve/kfmodel.py

```diff
@@ -108,40 +110,35 @@ def load(self) -> bool:
         self.ready = True
         return self.ready
 
-    async def preprocess(self, request: Dict) -> Union[Dict, ModelInferRequest]:
+    async def preprocess(self, request: Dict) -> Union[Dict, CloudEvent, ModelInferRequest]:
```

Comment:
I think if it is `CloudEvent` with protobuf/avro encoding, user should overwrite preprocess to decode with the corresponding data schema as KServe dose not know how to decode the message, so the return type should still be a `Dict`. The avro test case is a bit misleading, should move [`_parserequest`](https://github.com/kserve/kserve/blob/master/python/kserve/test/test_server.py#L136) into the `preprocess` handler instead of `predict`. Here we can change the request type to `Union[Dict, CloudEvent]` and leave return type as is.

Reply:
If users must overwrite preprocess for avro/protobuf, then do we still need this part?

https://github.com/kserve/kserve/blob/b02eb376271a6473ee04b449a64942e635372cc0/python/kserve/kserve/kfmodel.py#L129-L135

Reply:
In your code there is a slight behavior change, it now returns `request` instead of `request.data` for the binary payload case.

So we actually have something like following to deal with different payload format.
```python
def preprocess(self, inputs: Union[Dict, CloudEvent]) -> Dict:
        # We call preprocess form super class to return raw input. Its follows below interface.
        # if inputs is python dict, then output will be python dict
        # if inputs is CloudEvent containing msg in json format, then output will be python dict
        # if inputs is CloudEvent containing msg in raw bytes based format,
        # then output will be raw bytes
        inputs = super().preprocess(inputs)

        # if input is json/python dict, no decoding is required.
        # for raw bytes, msg will be decoded according to its schema

        if isinstance(inputs, bytes):            
            # Attempt parsing bytes as AVRO format
            parsed_avro_schema = avro.schema.parse(self.AVRO_SCHEMA)
            reader = avro.io.DatumReader(parsed_avro_schema)
            bytes_reader = io.BytesIO(inputs)
            decoder = avro.io.BinaryDecoder(bytes_reader)
            inputs = reader.read(decoder) 
```

Reply:
Ah I see now thanks 👍 

For now I will keep it the same so the CloudEvent changes can be merged. But I might revisit this later to see if I can make the process simpler/clearer.